rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,607 @@
+
+When any of the `target_health` fields change, the `db_server` heartbeat will be considered different from the old heartbeat, and the new health status will be announced as an updated `db_server`.
+
+When the database is deregistered from the agent, its health checker will be stopped.
+
+When the database is updated, it will be deregistered and then re-registered, effectively restarting its health checker and health status.
+
+DB agents will also watch `health_check_config` resources.
+
+When `health_check_config` resources are created, deleted, or updated, the DB agent will reevaluate the `health_check_config` label selectors for each registered database.
+
+If the health check settings for a database change, then its current health checker, if any, will be stopped, and a new health checker may be started.
+
+The database's health status will also be reset when its health check settings change.
+
+### Target address resolution
+
+The health check target address will be taken from the database uri.
+Teleport does not strictly enforce a host:port database uri, so the target host and/or port may need to be resolved.
+The target host and port should generally be resolved in the same way it is for user connections.
+
+As a special case, some databases have multiple endpoints.
+For example, a MongoDB replicaset connection string may look like this:
+
+    mongodb://host1:port1,host2:port2,host3:port3/?replicaSet=rs0
+
+To complicate matters further, the user may specify a "secondary" read preference, in which case operations will *only* read from secondary members of the replicaset and fail when secondaries are not available, even if the primary is available:
+
+    mongodb://host1:port1,host2:port2,host3:port3/?replicaSet=rs0&readPreference=secondary
+
+When there are multiple endpoints for a target, the health checker should check every endpoint in parallel and combine the results for each health check.
+The combined health check result is a failure if any of the endpoint checks fail.
+
+MongoDB also supports a `mongodb+srv://` scheme, which returns the replicaset "host:port" pairs from a DNS SRV record.
+In this case, the target endpoints can change dynamically based on the SRV record returned.
+Each health check should lookup the SRV record and check each endpoint returned.
+
+### Health checker behavior
+
+A health checker manager will be added that manages a mapping from DB name to a single health checker:
+
+    // Target is a health check target.
+    type Target struct {
+    	// Name is the uniquely identifying name of the target.
+    	Name string
+    	// GetEndpoints is callback func that returns the target endpoints.
+    	GetEndpoints func []string
+    	// Spec is the health check configuration to use for the target.
+    	Spec healthcheck.HealthCheckSpec
+    }
+
+    // Manager manages health checkers.
+    type Manager interface {
```

Comment:
how does manager handle a large number of databases? i think some agents may have hundreds or more databases?

Reply:
when agent restarts, will the agent try to connect all of databases at the same time?